pkg/ruler/manager.go

```diff
@@ -269,13 +306,25 @@ func (r *DefaultMultiTenantManager) getOrCreateNotifier(userID string, userManag
 	return n.notifier, nil
 }
 
+func (r *DefaultMultiTenantManager) getCachedRules(userID string) ([]*promRules.Group, bool) {
+	r.ruleCacheMtx.RLock()
+	defer r.ruleCacheMtx.RUnlock()
+	groups, exists := r.ruleCache[userID]
+	return groups, exists
+}
+
 func (r *DefaultMultiTenantManager) GetRules(userID string) []*promRules.Group {
 	var groups []*promRules.Group
-	r.userManagerMtx.Lock()
-	if mngr, exists := r.userManagers[userID]; exists {
+	groups, cached := r.getCachedRules(userID)
```

Comment:
I am trying to understand why we need to introduce rules caching in this pr. Can you please mention it in the PR description?

Is it because `manager.Update` during sync rules time can be slow and we want to minimize the impact of waiting on the shared mutex?

Reply:
you are correct! A single rule manager can still manage large number of rule groups and can take time to stop the rule groups and update the new ones. This will impact ListRules and we want to minimize the impact